node/src/Pos/Block/Logic/Header.hs

```diff
@@ -224,52 +223,52 @@ classifyHeaders inRecovery headers = do
-        mapM (DB.blkGetHeader @ssc) (toList checkpoints)
-    tip <- GS.getTip
-    unless (all ((/= tip) . headerHash) validCheckpoints) $
-        throwError "Found checkpoint that is equal to our tip"
-    let startFrom = fromMaybe tip startM
-        isCheckpoint bh =
-            any (\c -> bh ^. headerHashG == c ^. headerHashG) validCheckpoints
-        parentIsCheckpoint bh =
-            any (\c -> bh ^. prevBlockL == c ^. headerHashG) validCheckpoints
-        whileCond bh = not (isCheckpoint bh)
-    headers <- noteM "Failed to load headers by depth" . fmap (_Wrapped nonEmpty) $
-        DB.loadHeadersByDepthWhile whileCond recoveryHeadersMessage startFrom
-    let newestH = headers ^. _Wrapped . _neHead
-        oldestH = headers ^. _Wrapped . _neLast
-    logDebug $
-        sformat ("getHeadersFromManyTo: retrieved headers, oldest is "
-                % build % ", newest is " % build) oldestH newestH
-    if parentIsCheckpoint oldestH
-    then pure headers
-    else do
-        logDebug $ "getHeadersFromManyTo: giving headers in recovery mode"
-        inMainCheckpoints <-
-            noteM "Filtered set of valid checkpoints is empty" $
-            nonEmpty <$> filterM GS.isBlockInMainChain (toList validCheckpoints)
-        logDebug $ "getHeadersFromManyTo: got checkpoints in main chain"
-        let lowestCheckpoint =
-                maximumBy (comparing getEpochOrSlot) inMainCheckpoints
-            loadUpCond _ h = h < recoveryHeadersMessage
-        up <- GS.loadHeadersUpWhile lowestCheckpoint loadUpCond
-        res <- note "loadHeadersUpWhile returned empty list" $
-            _Wrapped nonEmpty (toNewestFirst $ over _Wrapped (drop 1) up)
-        logDebug $ "getHeadersFromManyTo: loaded non-empty list of headers, returning"
-        pure res
+
+    tip <- DB.getTipHeader @ssc
+    let tipHash = headerHash tip
+
+    -- This filters out invalid/unknown checkpoints also.
+    inMainCheckpoints <-
+        noteM "no checkpoints are in the main chain" $
+        nonEmpty <$> filterM GS.isBlockInMainChain (toList checkpoints)
+    let inMainCheckpointsHashes = map headerHash inMainCheckpoints
+    when (tipHash `elem` inMainCheckpointsHashes) $
+        throwError "found checkpoint that is equal to our tip"
+    logDebug $ "got checkpoints in main chain"
+
+    if (tip ^. prevBlockL . headerHashG) `elem` inMainCheckpointsHashes
+        -- Optimization for the popular case "just get me the newest
+        -- block, i know the previous one".
+        then pure $ one tip
+        -- If optimization doesn't apply, just iterate down-up
+        -- starting with the newest header.
+        else do
+            newestCheckpoint <-
+                maximumBy (comparing getEpochOrSlot) . catMaybes <$>
+                mapM (DB.blkGetHeader @ssc) (toList inMainCheckpoints)
+            let loadUpCond _ h = h < recoveryHeadersMessage
+            up <- GS.loadHeadersUpWhile @ssc newestCheckpoint loadUpCond
```

Comment:
As far as I can tell this should give a performance *regression* rather than improvement! Previously we would `loadHeadersByDepthWhile` in case one of the checkpoints is in the main chain, and that one is better because it doesn't have to deserialize any blocks. `loadHeadersUpWhile` apparently does, making it slower, although maybe we can patch it so that it doesn't have to deserialize blocks?

Reply:
> in case one of the checkpoints is in the main chain

This is not right. We did iterate `recoveryHeadersMessage` blocks deep or unless we meet a checkpoint, which was never a case with recovery, so we effectively throw one iteration away. Regarding `loadHeadersUpWhile` -- yes, it's stupid that the whole block is deserialized, i can easily optimize it. Thanks.